instrumentation/database/sql/splunksql/sql.go

```diff
@@ -0,0 +1,99 @@
+// Copyright Splunk Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package splunksql provides functions to trace the database/sql package
+// (https://golang.org/pkg/database/sql) using the OpenTelemetry API. It will
+// automatically augment operations such as connections, statements and
+// transactions with tracing.
+package splunksql // import "github.com/signalfx/splunk-otel-go/instrumentation/database/sql/splunksql"
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"sync"
+)
+
+var (
+	registryMu sync.RWMutex
+	registry   = make(map[string]InstrumentationConfig)
+)
+
+// Register makes the InstrumentationConfig for the setup of a database
+// driverwith the provided name available. If Register is called twice for the
+// same name it panics.
+func Register(name string, c InstrumentationConfig) {
+	registryMu.Lock()
+	defer registryMu.Unlock()
+	if _, dup := registry[name]; dup {
+		panic("splunksql: Register called twice for " + name)
+	}
+	registry[name] = c
+}
+
+func retrieve(name string) InstrumentationConfig {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+	return registry[name]
+}
+
+// Open opens a database specified by its database driver name and a
+// driver-specific data source name, usually consisting of at least a database
+// name and connection information. The returned database uses a traced driver
+// for all connections.
+func Open(driverName, dataSourceName string, opts ...Option) (*sql.DB, error) {
```

Comment:
What do you think of adding an `OpenDB` function? AFAIK there are devs who dislike the `sql.Open` b/c it uses a global state. I am aware it may be YAGNI, but prefer to ask just in case.

Reply:
The `OpenDB` function takes a `driver.Connector` and there is no way to get the needed connection information to properly annotate spans according to the OpenTelemetry semantic conventions. If we want to offer an `OpenDB` function it will need to be augmented from the function signature exported in the `database/sql` package. I want real user use cases before we do such a thing and leave that addition to the future.

Reply:
I guess it will require passing `InstrumentationConfig` and `...Option` right? What do you think of creating an GH issue for it? We can work on it if someone would e.g. upvote. I think it may be good to create an issue to leave the "design" proposal.

Reply:
Sure, feel free to create an issue if you think it would be useful :+1:. For what it's worth, it would need to take an `InstrumentationConfig` and a connection string, which would render it just a superficial superset of the existing `Open` which takes the driver name instead of connector. Otherwise it would need to be passed a Connection string, which would not be able to be validated as correct.